ghc 8.4: Can't get this to work

After fixing a few minor issues wrt. to `text-format` and `formatting` cross `universum` and using `o-clock` from GitHub, I keep running into:

```
src/Serokell/Util/Bench.hs:31:18: error:
        • Could not deduce (KnownRat
                              (Time.Rational.DivRat (1 Time.Rational.:% 1000000000) unit))
            arising from a use of ‘toUnit’
          from the context: KnownRat unit
            bound by the type signature for:
                       timeSpecToUnit :: forall (unit :: Time.Rational.Rat).
                                         KnownRat unit =>
                                         TimeSpec -> Time unit
            at src/Serokell/Util/Bench.hs:30:1-70
        • In the first argument of ‘(.)’, namely ‘toUnit @unit’
          In the expression: toUnit @unit . ns . fromIntegral . toNanoSecs
          In an equation for ‘timeSpecToUnit’:
              timeSpecToUnit = toUnit @unit . ns . fromIntegral . toNanoSecs
       |
    31 | timeSpecToUnit = toUnit @unit . ns . fromIntegral . toNanoSecs
       |                  ^^^^^^^^^^^^
```

with ghc-8.4, not so with ghc-8.2.

Now I've seen there are quite a few `#if ( __GLASGOW_HASKELL__ >= 804 )` conditionals in `o-clock`, but I'm unsure what the underlying reason might be. If anyone has seen this or can tell me how to fix it, I might be able to provide some PRs.

--- a/.travis.yml
+++ b/.travis.yml
@@ -15,8 +15,16 @@ matrix:
 
   include:
 
-  - ghc: 8.2.2
+  - ghc: 8.4.3
     env: STACK_YAML="$HOME/build/serokell/serokell-util/stack.yaml"
+    os: linux
+    addons:
+      apt:
+        packages:
+        - libgmp-dev
+
+  - ghc: 8.2.2
+    env: STACK_YAML="$HOME/build/serokell/serokell-util/stack-8.2.2.yaml"
     os: linux
     addons:
       apt:
@@ -31,15 +39,15 @@ matrix:
         packages:
         - libgmp-dev
 
-  - ghc: 8.2.2
-    env: GHCVER='8.2.2' CABALVER='head'
+  - ghc: 8.4.3
+    env: GHCVER='8.4.3' CABALVER='head'
     os: linux
     addons:
       apt:
         sources:
         - hvr-ghc
         packages:
-        - ghc-8.2.2
+        - ghc-8.4.3
         - cabal-install-head
 
 install:

--- a/CHANGES.md
+++ b/CHANGES.md
@@ -1,10 +1,15 @@
 0.10.0
 =====
 
+* [#67](https://github.com/serokell/serokell-util/issues/67):
+  GHC-8.4.3 is now officially supported. LTS-12.5 is used.
+  GHC-8.2.2 and GHC-8.0.2 are still supported as well.
+  Lower bounds on some dependencies (e. g. fmt, o-clock, universum)
+  have been increased.
 * [#65](https://github.com/serokell/serokell-util/issues/65):
   More recent version of `formatting` is now supported.
   `text-format` dependency was dropped, so there are no instances of
-      `Data.Text.Buildable` anymore.
+  `Data.Text.Buildable` anymore.
   Also some deprecated functions have been removed.
 
 0.9.0

--- a/serokell-util.cabal
+++ b/serokell-util.cabal
@@ -58,12 +58,12 @@ library
                      , deepseq
                      , exceptions
                      , fmt
-                     , formatting >= 6.3.4 && < 6.4
+                     , formatting ^>= 6.3.4
                      , hashable >= 1.2.4.0
                      , microlens
                      , microlens-mtl
                      , mtl
-                     , o-clock ^>= 0.1.1
+                     , o-clock ^>= 1.0.0
                      , parsec
                      , process
                      , QuickCheck >= 2.8.1
@@ -103,7 +103,7 @@ test-suite serokell-test
   build-depends:       aeson >= 1.0 && < 1.4
                      , base >=4.8
                      , extra >= 1.6
-                     , formatting
+                     , formatting ^>= 6.3.4
                      , hspec >= 2.1.10
                      , QuickCheck >= 2.8.1
                      , quickcheck-instances

--- a/serokell-util.cabal
+++ b/serokell-util.cabal
@@ -16,6 +16,9 @@ description:         Serokell-util is a library consisting of functions, which
                      functions from existing projects and don't want to
                      copy-paste them.
 cabal-version:       >=2.0
+tested-with:         GHC == 8.0.2
+                   , GHC == 8.2.2
+                   , GHC == 8.4.3
 
 library
   hs-source-dirs:      src

--- a/serokell-util.cabal
+++ b/serokell-util.cabal
@@ -73,7 +73,7 @@ library
                      , text
                      , th-lift-instances
                      , transformers
-                     , universum ^>= 1.1.0
+                     , universum ^>= 1.2.0
                      , unordered-containers >= 0.2.7.0
                      , vector
 
@@ -109,7 +109,7 @@ test-suite serokell-test
                      , quickcheck-instances
                      , scientific
                      , serokell-util
-                     , universum ^>= 1.1.0
+                     , universum ^>= 1.2.0
                      , unordered-containers >= 0.2.7.0
                      , vector
   hs-source-dirs:      test

--- a/src/Serokell/Util/Bench.hs
+++ b/src/Serokell/Util/Bench.hs
@@ -13,7 +13,7 @@ module Serokell.Util.Bench
        , perSecond
        ) where
 
-import Universum hiding (Buildable)
+import Universum
 
 import Fmt (Buildable (build), (+||), (||+))
 import System.Clock (Clock (..), TimeSpec, diffTimeSpec, getTime, toNanoSecs)

--- a/src/Serokell/Util/Bench.hs
+++ b/src/Serokell/Util/Bench.hs
@@ -1,5 +1,6 @@
-{-# LANGUAGE CPP                 #-}
 {-# LANGUAGE ExplicitForAll      #-}
+{-# LANGUAGE FlexibleContexts    #-}
+{-# LANGUAGE MonoLocalBinds      #-}
 {-# LANGUAGE ScopedTypeVariables #-}
 
 -- | Benchmark related utils.
@@ -17,17 +18,24 @@ import Universum
 
 import Fmt (Buildable (build), (+||), (||+))
 import System.Clock (Clock (..), TimeSpec, diffTimeSpec, getTime, toNanoSecs)
-import Time (KnownRat, Time, ns, toUnit)
+import Time (KnownDivRat, Nanosecond, Time, ns, toUnit)
 
 -- | Get current wall-clock time as any time unit.
-getWallTime :: forall unit m . (MonadIO m, KnownRat unit) => m (Time unit)
+getWallTime ::
+       forall unit m. (MonadIO m, KnownDivRat Nanosecond unit)
+    => m (Time unit)
 getWallTime = timeSpecToUnit <$> getTime' Realtime
 
 -- | Get current CPU time as any time unit.
-getCpuTime :: forall unit m . (MonadIO m, KnownRat unit) => m (Time unit)
+getCpuTime ::
+       forall unit m. (MonadIO m, KnownDivRat Nanosecond unit)
+    => m (Time unit)
 getCpuTime = timeSpecToUnit <$> getTime' ProcessCPUTime
 
-timeSpecToUnit :: forall unit . KnownRat unit => TimeSpec -> Time unit
+timeSpecToUnit ::
+       forall unit. KnownDivRat Nanosecond unit
+    => TimeSpec
+    -> Time unit
 timeSpecToUnit = toUnit @unit . ns . fromIntegral . toNanoSecs
 
 -- | Data type describing time passed during execution of something.

--- a/stack-8.0.2.yaml
+++ b/stack-8.0.2.yaml
@@ -11,6 +11,6 @@ extra-deps:
 - formatting-6.3.6
 - o-clock-0.1.1
 - recursion-schemes-5
-- universum-1.1.0
+- universum-1.2.0
 
 allow-newer: true

--- a/stack-8.0.2.yaml
+++ b/stack-8.0.2.yaml
@@ -7,9 +7,9 @@ extra-deps:
 - aeson-extra-0.4.0.0
 - aeson-options-0.0.0
 - containers-0.5.10.2
-- fmt-0.5.0.0
+- fmt-0.6
 - formatting-6.3.6
-- o-clock-0.1.1
+- o-clock-1.0.0
 - recursion-schemes-5
 - universum-1.2.0
 

--- a/stack-8.2.2.yaml
+++ b/stack-8.2.2.yaml
@@ -1,0 +1,10 @@
+resolver: lts-10.6
+
+nix:
+    packages: [openssh,git]
+
+extra-deps:
+- o-clock-1.0.0
+- fmt-0.6
+- universum-1.2.0
+- formatting-6.3.6

--- a/stack.yaml
+++ b/stack.yaml
@@ -1,9 +1,7 @@
-resolver: lts-10.6
+resolver: lts-12.5
 
 nix:
     packages: [openssh,git]
 
 extra-deps:
-- o-clock-0.1.1
-- universum-1.2.0
-- formatting-6.3.6
+- fmt-0.6

--- a/stack.yaml
+++ b/stack.yaml
@@ -5,5 +5,5 @@ nix:
 
 extra-deps:
 - o-clock-0.1.1
-- universum-1.1.0
+- universum-1.2.0
 - formatting-6.3.6

--- a/test/Test/Serokell/Util/TextSpec.hs
+++ b/test/Test/Serokell/Util/TextSpec.hs
@@ -7,6 +7,7 @@ module Test.Serokell.Util.TextSpec
 
 import Universum
 
+import Formatting (Buildable, build, sformat)
 import Test.Hspec (Spec, describe)
 import Test.Hspec.QuickCheck (prop)
 import Test.QuickCheck ((===))
@@ -31,4 +32,4 @@ spec =
                     \(a :: Word64)  -> Right a === showReadIntegral a
 
 showReadIntegral :: (Buildable a, Integral a) => a -> Either String a
-showReadIntegral = S.readDecimal . pretty
+showReadIntegral = S.readDecimal . sformat build

